### PR TITLE
Feature - Update to how API data for comments is served to users

### PR DESCRIPTION
**Pull request by nkirui**
**Goal**
Update to how API data for comments is served to users
- Update comment and replies relationships to match tree data structures.
- Update/Add endpoints to serve API data for the comments and replies as per spec.

### CODE REVIEW

File: backend/urls.py [hunk 1]
```diff
@@ -27,6 +27,7 @@
     path('poll/user/schedule', PollUserScheduleListAPI.as_view(), name='poll_user_schedule'),
     path('schema/', SpectacularAPIView.as_view(), name='schema'),
     path('schema/redoc/', SpectacularRedocView.as_view(url_name='api:schema'), name='redoc'),
+    path('comments/', include((comment_patterns, 'comment'))),
```
Comment (Gofven) [inline backend/urls.py]: Comment patterns already exist for respective module that uses it e.g. group/urls.py, poll/urls.py

File: flowback/comment/admin.py
Comment (Gofven) [inline flowback/comment/admin.py:1]: nice c:

File: flowback/comment/views.py [hunk 2]
```diff
@@ -4,11 +4,26 @@
 from rest_framework.views import APIView
 
 from flowback.comment.selectors import comment_list
-from flowback.comment.services import comment_create, comment_update, comment_delete
+from flowback.comment.services import comment_create, comment_delete, comment_update
 from flowback.common.pagination import LimitOffsetPagination, get_paginated_response
 from flowback.files.serializers import FileSerializer
 
 
+class CommentOutputSerializer(serializers.Serializer):
```
Comment (Gofven) [inline flowback/comment/views.py]: CommentOutputSerializer should either be used within OutputSerializer of an APIView, alternatively if used in more than one place, put it in comment/serializers.py

File: flowback/comment/views.py [hunk 3]
```diff
@@ -32,18 +47,28 @@ class FilterSerializer(serializers.Serializer):
...
-        author_name = serializers.CharField(source='author.username')
-        author_profile_image = serializers.ImageField(source='author.profile_image')
-        parent_id = serializers.IntegerField(allow_null=True)
-        created_at = serializers.DateTimeField()
-        edited = serializers.BooleanField()
-        active = serializers.BooleanField()
-        message = serializers.CharField()
-        attachments = FileSerializer(source="attachments.filesegment_set", many=True, allow_null=True)
-        score = serializers.IntegerField()
+    class OutputSerializer(CommentOutputSerializer):
```
Comment (Gofven) [inline flowback/comment/views.py]: Statement does nothing, delete or move CommentOutputSerializer contents to it,
If you delete it, change paginated_response serializer_class param to CommentOutputSerializer

**Review by Gofven — CHANGES_REQUESTED** [review]
Make sure to solve the merge conflicts before pushing this, also please add tests to see if these newly added features works as intended

Check the comments i've made, other than that it's looking quite good c:

File: flowback/comment/views.py
Comment (Gofven) [inline flowback/comment/views.py:1]: These APIViews are intended to be used elsewhere, don't hook these up with any urls directly, the point of lazy_action is for other modules to just override it with their own function there with permissions and such from respective services.py c:

File: backend/urls.py (hunk 1 above)
Comment (nkirui) [reply backend/urls.py]: I have removed the changes

File: flowback/comment/views.py
Comment (nkirui) [reply flowback/comment/views.py:1]: Well noted

Comment (nkirui) [reply flowback/comment/views.py:1]: Reverted the change in urls

File: flowback/comment/views.py (hunk 2 above)
Comment (nkirui) [reply flowback/comment/views.py]: I have moved it to a new serializers.py file and reused it the same way

File: flowback/comment/views.py (hunk 3 above)
Comment (nkirui) [reply flowback/comment/views.py]: Removed and updated the argument to use the new import

**Comment by nkirui** [comment]
> Make sure to solve the merge conflicts before pushing this, also please add tests to see if these newly added features works as intended
> 
> Check the comments i've made, other than that it's looking quite good c:

Resolved the conflicts and added tests to the group app for adding comments to threads

**Comment by nkirui** [comment]
cc @Gofven

File: flowback/group/services.py [hunk 4]
```diff
@@ -516,11 +516,7 @@ def group_thread_delete(user_id: int, thread_id: int):
     thread.delete()
 
 
-def group_thread_comment_create(author_id: int,
-                                thread_id: int,
-                                message: str = None,
-                                attachments: list = None,
-                                parent_id: int = None):
+def group_thread_comment_create(*,author_id, thread_id, parent_id, message,attachments,**kwargs):
```
Comment (Gofven) [inline flowback/group/services.py]: It'd be better having proper typehints to help any future maintainers of code, including yourself

Additionally people should be able to send either message or attachments

Also kwargs have no place in here

File: flowback/group/tests/test_group.py [hunk 5]
```diff
@@ -161,3 +167,124 @@ def test_user_group_join_request_list(self):
...
+        self.assertEqual(len(response.data["results"]), 10)
+        self.assertEqual([msg["id"] for msg in response.data["results"]], comment_ids)
+        self.assertTrue(extra_resp.data not in [msg["id"] for msg in response.data["results"]])
+
+    def test_can_fetch_ancestors_of_a_thread_comment(self):
+        create_view = GroupThreadCommentCreateAPI.as_view()
+        parent_id = None
+        parent_comment_id = None
+        comment_ids = []
+        for i in range(10):
```
Comment (Gofven) [inline flowback/group/tests/test_group.py:263]: In the future, you can use factories.py to produce multiple comments much easier c:

**Review by Gofven — COMMENTED** [review]
Not bad! Although it's missing poll comments, and a way to view all comments in a feed through a tree-like structure!

**Comment by nkirui** [comment]
@Gofven was thinking of excluding pagination when descendants/ancestors are requested. I let the ancestors/descendant requests return a list to easily support pagination. To return a complex tree like object I would like to do away with pagination. WDYT?

File: flowback/group/services.py (hunk 4 above)
Comment (nkirui) [reply flowback/group/services.py]: Updated to have type hints and remove the kwargs

**Pull request closed without merging**